### PR TITLE
[#16] 이메일 인증 기능

### PR DESCRIPTION
**Pull request by JinDDung2**
## 작업 요약

- 회원가입 시 필요한 이메일 인증

## 작업 내용

- [x] 인증코드를 위한 난수 생성
- [x] 인증코드 레디스에 저장
- [x] 인증코드 일치 확인
- [x] 이메일을 찾지 못할 경우, 인증코드가 일치하지 않은 경우 예외처리

### CODE REVIEW

**Comment by JinDDung2** [comment]
### 📝 테스트 커버리지 리포트입니다
|File|Coverage [90.78%]|:green_apple:|
|:-|:-:|:-:|
|[ErrorCode.java](https://github.com/f-lab-edu/givemeticon/blob/a31453fc8a4ee4ce7364b576c570b425589be6e3/src%2Fmain%2Fjava%2Fcom%2Fjinddung2%2Fgivemeticon%2Fcommon%2Fexception%2FErrorCode.java)|100%|:green_apple:|
|[MailVerifyService.java](https://github.com/f-lab-edu/givemeticon/blob/a31453fc8a4ee4ce7364b576c570b425589be6e3/src%2Fmain%2Fjava%2Fcom%2Fjinddung2%2Fgivemeticon%2Fmail%2Fapplication%2FMailVerifyService.java)|100%|:green_apple:|
|[MailSendService.java](https://github.com/f-lab-edu/givemeticon/blob/a31453fc8a4ee4ce7364b576c570b425589be6e3/src%2Fmain%2Fjava%2Fcom%2Fjinddung2%2Fgivemeticon%2Fmail%2Fapplication%2FMailSendService.java)|100%|:green_apple:|
|[MailController.java](https://github.com/f-lab-edu/givemeticon/blob/a31453fc8a4ee4ce7364b576c570b425589be6e3/src%2Fmain%2Fjava%2Fcom%2Fjinddung2%2Fgivemeticon%2Fmail%2Fpresentation%2FMailController.java)|100%|:green_apple:|
|[MailExceptionAdvice.java](https://github.com/f-lab-edu/givemeticon/blob/a31453fc8a4ee4ce7364b576c570b425589be6e3/src%2Fmain%2Fjava%2Fcom%2Fjinddung2%2Fgivemeticon%2Fmail%2Fadvice%2FMailExceptionAdvice.java)|30.43%|:x:|
|[EmailConstants.java](https://github.com/f-lab-edu/givemeticon/blob/a31453fc8a4ee4ce7364b576c570b425589be6e3/src%2Fmain%2Fjava%2Fcom%2Fjinddung2%2Fgivemeticon%2Fmail%2Fconstans%2FEmailConstants.java)|0%|:x:|

|Total Project Coverage|78.96%|:green_apple:|
|:-|:-:|:-:|

File: src/main/java/com/jinddung2/givemeticon/mail/application/MailVerifyService.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    private final CertificationNumberDao certificationNumberDao;
+
+    public void verifyEmail(String email, String certificationNumber) {
+        if (!isVerify(email, certificationNumber)) {
+            throw new InvalidCertificationNumberException();
+        }
+        certificationNumberDao.removeCertificationNumber(email);
+    }
+
+    private boolean isVerify(String email, String certificationNumber) {
```
Comment (JinDDung2) [inline src/main/java/com/jinddung2/givemeticon/mail/application/MailVerifyService.java:22]: 이메일 존재 확인, 인증코드 불일치 두 가지 예외를 위해 메서드를 분리했는데, 적절한 조치인지 궁금합니다.

**Comment by JinDDung2** [comment]
### 📝 테스트 커버리지 리포트입니다
|File|Coverage [90.82%]|:green_apple:|
|:-|:-:|:-:|
|[ErrorCode.java](https://github.com/f-lab-edu/givemeticon/blob/38e1a607211589d990116441c7c07d199ac5f9c7/src%2Fmain%2Fjava%2Fcom%2Fjinddung2%2Fgivemeticon%2Fcommon%2Fexception%2FErrorCode.java)|100%|:green_apple:|
|[MailVerifyService.java](https://github.com/f-lab-edu/givemeticon/blob/38e1a607211589d990116441c7c07d199ac5f9c7/src%2Fmain%2Fjava%2Fcom%2Fjinddung2%2Fgivemeticon%2Fmail%2Fapplication%2FMailVerifyService.java)|100%|:green_apple:|
|[MailSendService.java](https://github.com/f-lab-edu/givemeticon/blob/38e1a607211589d990116441c7c07d199ac5f9c7/src%2Fmain%2Fjava%2Fcom%2Fjinddung2%2Fgivemeticon%2Fmail%2Fapplication%2FMailSendService.java)|100%|:green_apple:|
|[MailController.java](https://github.com/f-lab-edu/givemeticon/blob/38e1a607211589d990116441c7c07d199ac5f9c7/src%2Fmain%2Fjava%2Fcom%2Fjinddung2%2Fgivemeticon%2Fmail%2Fpresentation%2FMailController.java)|100%|:green_apple:|
|[MailExceptionAdvice.java](https://github.com/f-lab-edu/givemeticon/blob/38e1a607211589d990116441c7c07d199ac5f9c7/src%2Fmain%2Fjava%2Fcom%2Fjinddung2%2Fgivemeticon%2Fmail%2Fadvice%2FMailExceptionAdvice.java)|30.43%|:x:|
|[EmailConstants.java](https://github.com/f-lab-edu/givemeticon/blob/38e1a607211589d990116441c7c07d199ac5f9c7/src%2Fmain%2Fjava%2Fcom%2Fjinddung2%2Fgivemeticon%2Fmail%2Fconstans%2FEmailConstants.java)|0%|:x:|

|Total Project Coverage|79.01%|:green_apple:|
|:-|:-:|:-:|

File: src/main/java/com/jinddung2/givemeticon/mail/advice/MailExceptionAdvice.java [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+import org.springframework.web.bind.annotation.RestControllerAdvice;
+
+@RestControllerAdvice
+@Slf4j
+public class MailExceptionAdvice {
+
+    @ExceptionHandler(MailException.class)
+    public ResponseEntity<ErrorResult> handleMailException(MailException e) {
+        ErrorResult errorResult = new ErrorResult(e.getMessage());
+        log.info("mail exception!! error msg={}", errorResult);
```
Comment (f-lab-kyle) [inline src/main/java/com/jinddung2/givemeticon/mail/advice/MailExceptionAdvice.java]: 로그레벨이 고민되네요

File: src/main/java/com/jinddung2/givemeticon/mail/constans/EmailConstants.java [hunk 3]
```diff
@@ -0,0 +1,7 @@
+package com.jinddung2.givemeticon.mail.constans;
+
+public class EmailConstants {
+    public static final String MAIL_TITLE_CERTIFICATION = "givemeticon 인증번호 안내";
+    public static final String DOMAIN_NAME = "http://localhost:8080";
```
Comment (f-lab-kyle) [inline src/main/java/com/jinddung2/givemeticon/mail/constans/EmailConstants.java]: 요런 값은 프로퍼티로 관리하는게 맞을것 같아요. 환경마다 달라질테니

File: src/main/java/com/jinddung2/givemeticon/mail/application/MailSendService.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    private final CertificationNumberDao certificationNumberDao;
+    private final CertificationGenerator generator;
+
+    public EmailCertificationResponse sendEmailForCertification(String email) throws NoSuchAlgorithmException, MessagingException {
+
+        String certificationNumber = generator.createCertificationNumber();
+
+        String content = String.format("%s/api/v1/users/verify?certificationNumber=%s&email=%s   링크를 3분 이내에 클릭해주세요.", DOMAIN_NAME, certificationNumber, email);
+        sendMail(email, content);
+        certificationNumberDao.saveCertificationNumber(email, certificationNumber);
```
Comment (f-lab-kyle) [inline src/main/java/com/jinddung2/givemeticon/mail/application/MailSendService.java]: 순서가 바껴야 하지 않을까요?

**Review by f-lab-kyle — COMMENTED** [review]
테스트의 이름은 항상 상황에 따른 기대값에 대한 서술형으로 작성하는 편입니다~
수고하셨습니다~

**Comment by JinDDung2** [comment]
### 📝 테스트 커버리지 리포트입니다
|File|Coverage [92.31%]|:green_apple:|
|:-|:-:|:-:|
|[ErrorCode.java](https://github.com/f-lab-edu/givemeticon/blob/05b66b770b68dcf11670e02cf510cb2c87f6fbb7/src%2Fmain%2Fjava%2Fcom%2Fjinddung2%2Fgivemeticon%2Fcommon%2Fexception%2FErrorCode.java)|100%|:green_apple:|
|[MailVerifyService.java](https://github.com/f-lab-edu/givemeticon/blob/05b66b770b68dcf11670e02cf510cb2c87f6fbb7/src%2Fmain%2Fjava%2Fcom%2Fjinddung2%2Fgivemeticon%2Fmail%2Fapplication%2FMailVerifyService.java)|100%|:green_apple:|
|[MailSendService.java](https://github.com/f-lab-edu/givemeticon/blob/05b66b770b68dcf11670e02cf510cb2c87f6fbb7/src%2Fmain%2Fjava%2Fcom%2Fjinddung2%2Fgivemeticon%2Fmail%2Fapplication%2FMailSendService.java)|100%|:green_apple:|
|[MailController.java](https://github.com/f-lab-edu/givemeticon/blob/05b66b770b68dcf11670e02cf510cb2c87f6fbb7/src%2Fmain%2Fjava%2Fcom%2Fjinddung2%2Fgivemeticon%2Fmail%2Fpresentation%2FMailController.java)|100%|:green_apple:|
|[MailExceptionAdvice.java](https://github.com/f-lab-edu/givemeticon/blob/05b66b770b68dcf11670e02cf510cb2c87f6fbb7/src%2Fmain%2Fjava%2Fcom%2Fjinddung2%2Fgivemeticon%2Fmail%2Fadvice%2FMailExceptionAdvice.java)|30.43%|:x:|

|Total Project Coverage|79.8%|:green_apple:|
|:-|:-:|:-:|

**Comment by JinDDung2** [comment]
### 📝 테스트 커버리지 리포트입니다
|File|Coverage [100%]|:green_apple:|
|:-|:-:|:-:|
|[ErrorCode.java](https://github.com/f-lab-edu/givemeticon/blob/a764d4006d278d6e62c93cfbe50877be85e23908/src%2Fmain%2Fjava%2Fcom%2Fjinddung2%2Fgivemeticon%2Fcommon%2Fexception%2FErrorCode.java)|100%|:green_apple:|
|[MailVerifyService.java](https://github.com/f-lab-edu/givemeticon/blob/a764d4006d278d6e62c93cfbe50877be85e23908/src%2Fmain%2Fjava%2Fcom%2Fjinddung2%2Fgivemeticon%2Fmail%2Fapplication%2FMailVerifyService.java)|100%|:green_apple:|
|[MailSendService.java](https://github.com/f-lab-edu/givemeticon/blob/a764d4006d278d6e62c93cfbe50877be85e23908/src%2Fmain%2Fjava%2Fcom%2Fjinddung2%2Fgivemeticon%2Fmail%2Fapplication%2FMailSendService.java)|100%|:green_apple:|
|[MailExceptionAdvice.java](https://github.com/f-lab-edu/givemeticon/blob/a764d4006d278d6e62c93cfbe50877be85e23908/src%2Fmain%2Fjava%2Fcom%2Fjinddung2%2Fgivemeticon%2Fmail%2Fadvice%2FMailExceptionAdvice.java)|100%|:green_apple:|
|[MailController.java](https://github.com/f-lab-edu/givemeticon/blob/a764d4006d278d6e62c93cfbe50877be85e23908/src%2Fmain%2Fjava%2Fcom%2Fjinddung2%2Fgivemeticon%2Fmail%2Fpresentation%2FMailController.java)|100%|:green_apple:|

|Total Project Coverage|85.78%|:green_apple:|
|:-|:-:|:-:|

**Comment by JinDDung2** [comment]
### 📝 테스트 커버리지 리포트입니다
|File|Coverage [100%]|:green_apple:|
|:-|:-:|:-:|
|[ErrorCode.java](https://github.com/f-lab-edu/givemeticon/blob/a764d4006d278d6e62c93cfbe50877be85e23908/src%2Fmain%2Fjava%2Fcom%2Fjinddung2%2Fgivemeticon%2Fcommon%2Fexception%2FErrorCode.java)|100%|:green_apple:|
|[MailVerifyService.java](https://github.com/f-lab-edu/givemeticon/blob/a764d4006d278d6e62c93cfbe50877be85e23908/src%2Fmain%2Fjava%2Fcom%2Fjinddung2%2Fgivemeticon%2Fmail%2Fapplication%2FMailVerifyService.java)|100%|:green_apple:|
|[MailSendService.java](https://github.com/f-lab-edu/givemeticon/blob/a764d4006d278d6e62c93cfbe50877be85e23908/src%2Fmain%2Fjava%2Fcom%2Fjinddung2%2Fgivemeticon%2Fmail%2Fapplication%2FMailSendService.java)|100%|:green_apple:|
|[MailExceptionAdvice.java](https://github.com/f-lab-edu/givemeticon/blob/a764d4006d278d6e62c93cfbe50877be85e23908/src%2Fmain%2Fjava%2Fcom%2Fjinddung2%2Fgivemeticon%2Fmail%2Fadvice%2FMailExceptionAdvice.java)|100%|:green_apple:|
|[MailController.java](https://github.com/f-lab-edu/givemeticon/blob/a764d4006d278d6e62c93cfbe50877be85e23908/src%2Fmain%2Fjava%2Fcom%2Fjinddung2%2Fgivemeticon%2Fmail%2Fpresentation%2FMailController.java)|100%|:green_apple:|

|Total Project Coverage|85.78%|:green_apple:|
|:-|:-:|:-:|

**Pull request merged**